README.md
# Forestry Log

[![Latest Version](https://img.shields.io/github/release/ForestryCodes/log.svg?style=flat-square)](https://github.com/ForestryCodes/log/releases)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/ForestryCodes/log/master.svg?style=flat-square)](https://travis-ci.org/ForestryCodes/log)
[![Codacy Badge](https://www.codacy.com/project/badge/9be442da34e34e548af5312f844fc0fe)](https://www.codacy.com/public/danielgithub/log)
[![Total Downloads](https://img.shields.io/packagist/dt/forestry/log.svg?style=flat-square)](https://packagist.org/packages/forestry/log)

Small log file writer with adjustable log level threshold settings.

## Install

Via Composer

``` bash
$ composer require forestry/log
```

## Usage

### Creating a logger

```php
$logger = new Forestry\Log\Log('/tmp/dummy.log');
//Just log notices and above.
$errorLog = new Forestry\Log\Log('./logs/error.log', Psr\Log\LogLevel::NOTICE);
```

#### Using a factory

Another way to create an instance is the use of one of the factories. There is one for each threshold level.

Here is an example for a logger with an error threshold:

```php
$factory = new ErrorLogger();
$logger = $factory->create('/tmp/error.log');
```

### Log a message

Forestry\Log provides methods for the log levels defined by RFC 5424 (debug, info, notice, warning, error, critical, alert and emergency). There's a method for each of these levels:

```php
$logger->emergency('This is an emergency message');
$logger->alert('This is an alert message');
$logger->critical('This is an critical message');
$logger->error('This is an error message');
$logger->warning('This is an warning message');
$logger->notice('This is a notice message');
$logger->info('This is an information');
$logger->debug('This is a debug message');
```

You can also use a generic `log` method:

```php
$logger->log(Psr\Log\LogLevel::DEBUG, 'this is a debug message');
```

#### Using placeholders in log messages

You can use placeholders in the your message string and fill them using the associative context array. The array keys have to match the placeholders without the curly brackets:

```php
$user = array('name' => 'John Doe', 'mail' => 'j.doe@example.org');
$logger->info('Send mail to {name} ({mail})', $user); //Send mail to John Doe (j.doe@example.org)
```

### Change the date format

The default date format is `Y-m-d H:i:s`. You can change it by using the `setDateFormat` method:

```php
$logger->setDateFormat('r'); //e.g. Thu, 21 Dec 2000 16:01:07 +0200
```

This method accepts any string which is compatible with PHPs native `date()`.

### Change the message format

The default format for the log message is `date level message`. To change it, you can re-arrange the placeholders with `setLogFormat`:

```php
$logger->setLogFormat('[{level}|{date}] {message}'); //[INFO|2013-04-25 13:37:42] This is an info message
```

There are the following placeholder available:

* {date}
* {level}
* {message}

These placeholders will be replaced with ones for `sprintf()`, so you can also use the following:

* `%1$s` = date
* `%2$s` = level
* `%3$s` = message

### Change the threshold level of an existing instance

To change the threshold level of an existing instance, use the `setLogThreshold` method:

```php
$logger->setLogThreshold(Psr\Log\LogLevel::DEBUG);
```

### Get current logging level

To get the current threshold level of an existing instance, use the `getLogThreshold` method:

```php
$level = $logger->getLogThreshold();
$logger->setLogThreshold(Psr\Log\LogLevel::INFO);
$logger->logInfo('my info');
$logger->setLogThreshold($level);
```

## Testing

``` bash
$ phpunit
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Credits

- [daniel-melzer](https://github.com/daniel-melzer)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All Notable changes to `log` will be documented in this file.

## 1.0.0 - 2015-05-02

### Added
- Added DirectoryException, FileException and own InvalidArgumentException.
- Added factories for each log level.

### Changed
- Changed parameter of the constructor. File path and name are not separated anymore.

## 0.3.0 - 2015-02-13

Implementation of PSR-3 introducing breaking changes.

### Added
- Psr\Log (PSR-3) implementation
- setLogThreshold() and getLogThreshold() methods

### Changed
- Reformat code according to the PSR-2 specs

### Fixed
- Fixed instructions and examples in README.md

### Removed
- Removed setLevel() and getLevel() methods

## 0.2.0.1 - 2015-02-02

No functionality added. Changes in the project structure and namespaces.

## 0.2.0 - 2014-09-23

### Removed
- Support for PHP 5.3

## 0.1.0 - 2013-06-05

First release.

src/Log.php
<?php
/**
 * This file is part of the Forestry Log library
 *
 * @license http://opensource.org/licenses/MIT
 * @link https://github.com/ForestryCodes/log/
 * @package Forestry Log
 *
 * For the full copyright and license information, please view the LICENSE file
 * that was distributed with this source code.
 */

namespace Forestry\Log;

use Psr\Log\AbstractLogger;
use Psr\Log\LogLevel;

/**
 * Class to write log files
 *
 * @package Forestry Log
 */
class Log extends AbstractLogger
{
    /**
     * Handle for all file operations.
     *
     * @var resource
     */
    private $handle;

    /**
     * Format of the date information.
     *
     * @var string
     */
    private $dateFormat = 'Y-m-d H:i:s';

    /**
     * Line format of a log entry.
     *
     * @var string
     */
    private $logFormat = '%1$s %2$s %3$s';

    /**
     * Severity values for each log level.
     *
     * @var array
     */
    private $levelSeverity = [
        LogLevel::EMERGENCY => 80,
        LogLevel::ALERT => 70,
        LogLevel::CRITICAL => 60,
        LogLevel::ERROR => 50,
        LogLevel::WARNING => 40,
        LogLevel::NOTICE => 30,
        LogLevel::INFO => 20,
        LogLevel::DEBUG => 10
    ];

    /**
     * Threshold value of the current Log instance.
     *
     * @var integer
     */
    private $thresholdLevel;

    /**
     * Constructor method.
     *
     * @param string $fileName
     * @param string $threshold
     * @throws DirectoryException
     * @throws FileException
     */
    public function __construct($fileName, $threshold = LogLevel::DEBUG)
    {
        $folder = dirname($fileName);
        if (!is_dir($folder) || !is_writable($folder)) {
            throw new DirectoryException('Folder does not exist, or is not writable');
        }

        $this->setLogThreshold($threshold);
        $this->handle = fopen($fileName, 'a');

        if (!$this->handle) {
            throw new FileException('Error opening log file with path: ' . $fileName);
        }
    }

    /**
     * Closes the handle.
     */
    public function __destruct()
    {
        if ($this->handle) {
            fclose($this->handle);
        }
    }

    /**
     * Writes a log message of a given level.
     *
     * @param string $level
     * @param string $message
     * @param array $context
     * @return null
     * @throws InvalidArgumentException
     */
    public function log($level, $message, array $context = array())
    {
        if (!isset($this->levelSeverity[$level])) {
            throw new InvalidArgumentException('invalid log level ' . $level);
        }

        if ($this->thresholdLevel <= $this->levelSeverity[$level]) {
            $message = sprintf(
                $this->logFormat . PHP_EOL,
                date($this->dateFormat),
                strtoupper($level),
                $this->interpolate((string)$message, $context)
            );

            fwrite($this->handle, $message);
        }
    }

    /**
     * Sets the date format.
     *
     * Accepts any string which is compatible with date().
     *
     * @param string $format
     * @return null
     */
    public function setDateFormat($format)
    {
        $this->dateFormat = $format;
    }

    /**
     * Sets the log string format.
     *
     * String will be parsed with sprintf(). Following placeholder are defined:
     * - {date}
     * - {level}
     * - {message}
     *
     * Example:
     * $format = '[{level}|{date}] {message}'; //[INFO|2013-04-25 13:37:42] This is an information message
     *
     * @param $format
     * @return null
     */
    public function setLogFormat($format)
    {
        $this->logFormat = str_replace(
            ['{date}', '{level}', '{message}'],
            ['%1$s', '%2$s', '%3$s'],
            $format
        );
    }

    /**
     * Sets the threshold level of the log instance.
     *
     * @param integer $level the new level to set
     * @return null
     * @throws InvalidArgumentException
     */
    public function setLogThreshold($level)
    {
        if (!isset($this->levelSeverity[$level])) {
            throw new InvalidArgumentException('invalid log level ' . $level);
        }

        $this->thresholdLevel = $this->levelSeverity[$level];
    }

    /**
     * Gets the current threshold level of the log instance.
     *
     * @return string
     */
    public function getLogThreshold()
    {
        return array_search($this->thresholdLevel, $this->levelSeverity);
    }

    /**
     * Replaces placeholders in the message with the values from the context array.
     *
     * @param string $message
     * @param array $context
     * @return string
     */
    private function interpolate($message, array $context = array())
    {
        $replace = [];
        foreach ($context as $key => $value) {
            $replace['{' . $key . '}'] = $value;
        }

        return strtr($message, $replace);
    }
}
